pkg/apis/core/validation/execution.go

```diff
@@ -40,21 +30,122 @@ func ValidateExecutionSpec(fldpath *field.Path, spec core.ExecutionSpec) field.E
 func ValidateDeployItemTemplateList(fldPath *field.Path, list core.DeployItemTemplateList) field.ErrorList {
 	allErrs := field.ErrorList{}
 	names := sets.NewString()
+	hasDuplicates := false
 	for i, tmpl := range list {
 		tmplPath := fldPath.Index(i)
 		if len(tmpl.Name) != 0 {
 			if names.Has(tmpl.Name) {
 				allErrs = append(allErrs, field.Duplicate(tmplPath, tmpl.Name))
+				hasDuplicates = true
 			}
 			names.Insert(tmpl.Name)
 			tmplPath = tmplPath.Key(tmpl.Name)
 		}
 		allErrs = append(allErrs, ValidateDeployItemTemplate(tmplPath, tmpl)...)
 	}
 
+	if !hasDuplicates { // cycle check identifies items by name and the behaviour is undefined if duplicate items are present
+		done := sets.NewString()
+		cycles := []Cycle{}
+		undefined := []UndefinedDeployItemReference{}
+		for i := range list {
+			c, u, d := getCyclesAndUndefinedDependencies(list, i, []string{}, done)
+			cycles = append(cycles, c...)
+			undefined = append(undefined, u...)
+			done.Insert(d.UnsortedList()...)
+		}
+		for _, c := range cycles {
+			allErrs = append(allErrs, field.Invalid(fldPath, c, "cycle found in dependencies"))
+		}
+		for _, u := range undefined {
+			idx, _ := getDeployItemTemplateByName(list, u.Source)
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(idx).Key(u.Source), u.Target, "depends on undefined deploy item"))
+		}
+	}
+
 	return allErrs
 }
 
+// takes arguments:
+// 1. a list of deploy item templates
+// 2. the index of the deploy item template that should be validated
+// 3. a list of already visited deploy item templates (referenced by name), shows the current state of the depth-first search
+// 4. a set of deploy item templates (referenced by name) that have already been checked and can be ignored
+// returns:
+// 1. a list of Cycle objects, representing found cyclic dependencies
+// 2. a list of UndefinedDeployItemReference objects, representing found dependencies to undefined deploy items
+// 3. a set of all deploy item templates (referenced by name) that have already been checked (necessary to avoid finding the same cycle multiple times)
+func getCyclesAndUndefinedDependencies(list core.DeployItemTemplateList, index int, visited []string, done sets.String) ([]Cycle, []UndefinedDeployItemReference, sets.String) {
```

Comment:
why do you need the list and index, isnt the deployitem enough?

maybe it would make sense to convert the deployitems list into map of `map[<name>]DeployItemTemplate`.
With that I think we could simplify the code here a lot.
WDYT?

Reply:
Since I'm basically doing a depth-first search along a graph, I need a method that can see the whole graph (= the list of all deploy items), since they are not linked within each other but just referenced by name. If this method is changed to just get one deploy item, a second method would be needed that knows the whole list and it would likely become more instead of less complicated.

I also thought about converting the list into a linked-list-style graph structure first, but I discarded that idea because it doesn't really remove more complexity than it adds.

Converting the list into the map you mentioned is quite easy, we could do that. The advantage is that we can get rid of the `getDeployItemTemplateByName` method. There is a small disadvantage, though: we lose the information about the index, which is currently added to the validation error in case of references to undefined deploy items. But it's anyhow easier to find them by name, so that shouldn't be a problem, I guess. So, should we implement this change?

Reply:
> Converting the list into the map you mentioned is quite easy, we could do that. The advantage is that we can get rid of the getDeployItemTemplateByName method. There is a small disadvantage, though: we lose the information about the index
We could co0nvert the list of deploy items to a map of `map[string]int` with that we sould not loose any information?

Reply:
If I understand you correctly, you would add a map that maps deploy item names to their corresponding index in the list?
With this we could also get rid of the `getDeployItemTemplateByName` method. This approach doesn't lose information, but we would need this map in addition to the already present list - adding another argument to the `getCyclesAndUndefinedDependencies` method - and it would contain redundant information.

Another option would be to transform the list into a map that maps the name to a custom struct containing the index as well as the deploy item template.

I think there is not really a nice solution. We would need some kind of hybrid data structure, that preserves the given order but still allows to index by a specific field ...